src/pages/announcements/utils.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2024, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import yup from '../../utils/yup-config';
+import { Duration, parse, serialize } from 'tinyduration';
+
+export const ID = 'id';
+export const DATE = 'creationDate';
+export const MESSAGE = 'message';
+export const DURATION = 'duration';
+export const DAYS = 'days';
+export const HOURS = 'hours';
+export const MINUTES = 'minutes';
+
+export interface DurationFormData {
+    [DAYS]: number | null;
+    [HOURS]: number | null;
+    [MINUTES]: number | null;
+}
+
+export interface AnnouncementFormData {
+    [MESSAGE]: string;
+    [DURATION]: DurationFormData;
+}
+
+export const emptyFormData: AnnouncementFormData = {
+    [MESSAGE]: '',
+    [DURATION]: {
+        [DAYS]: null,
+        [HOURS]: null,
+        [MINUTES]: null,
+    },
+};
+
+// const formSchema: ObjectSchema<FormData> = yup
+export const formSchema = yup
+    .object()
+    .shape({
+        [MESSAGE]: yup.string().max(100).required(),
+        [DURATION]: yup
+            .object()
+            .shape(
+                {
+                    [DAYS]: getDurationUnitSchema(HOURS, MINUTES),
+                    [HOURS]: getDurationUnitSchema(DAYS, MINUTES),
+                    [MINUTES]: getDurationUnitSchema(DAYS, HOURS),
+                },
+                // to avoid cyclic dependencies
+                [
+                    [HOURS, MINUTES],
+                    [DAYS, MINUTES],
+                    [DAYS, HOURS],
+                ]
+            )
+            .required(),
+    })
+    .required();
+
+function getDurationUnitSchema(otherUnitName1: string, otherUnitName2: string) {
+    return yup.number().when([otherUnitName1, otherUnitName2], {
+        is: (v1: number | null, v2: number | null) =>
+            v1 === null && v2 === null,
+        then: (schema) => schema.required(),
+        otherwise: (schema) => schema.nullable(),
+    });
+}
```

Comment:
Can't we simplify by saying at least one of these three is required ?

Reply:
I don't see how. We can do that for the validation but then the fields won't be red with "required" key word.